### PR TITLE
T15725

### PR DESCRIPTION
**Pull request by tchx84**
https://phabricator.endlessm.com/T15725

### CODE REVIEW

**Comment by tchx84** [comment]
NOTE FOR REVIWER: this PR depends on https://github.com/tchx84/basin/pull/1

File: Makefile.am [hunk 1]
```diff
@@ -18,7 +18,10 @@ EXTRA_DIST =
 # # # BUILD RULES # # #
 
 # Launcher script
-CLEANFILES += basin
+CLEANFILES += \
```
Comment (ptomato) [inline Makefile.am]: This is duplicated on lines 59 and following.

File: src/packer.js [hunk 2]
```diff
@@ -91,7 +91,7 @@ const Packer = new Lang.Class ({
     },
 
     _add_set: function (metadata) {
-        let set_hash = this._hashify(metadata['title']);
+        let set_hash = this._hashify(metadata['source']);
```
Comment (ptomato) [inline src/packer.js]: Unrelated change?

File: tools/generator.js [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+                this._sequence += 1;
+
+                let category = null;
+                if (categories.length > 0)
+                    category = categories[categories.length - 1];
+
+                output['articles'].push(this._create_article(path, name, mime_type, category, this._sequence));
+
+            } else {
+               /* mehhh */
```
Comment (ptomato) [inline tools/generator.js]: Print an error message?

File: tools/generator.js [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+                let thumbnail_path = path + '/thumbnail.jpg';
+
+                output['sets'].push(this._create_set(path, name, [name], tags, thumbnail_path));
+
+            } else if (mime_type.startsWith('image')) {
+                output['images'].push(this._create_image(path, mime_type));
+
+            } else if (mime_type === 'application/pdf') {
+                this._sequence += 1;
```
Comment (ptomato) [inline tools/generator.js]: Maybe something for a subsequent iteration, but it would be nice if you only got sequence numbers when you asked for them specifically (and then they were assigned in alphabetical order so that you could name your files `00-something`, `01-something`, etc.) Without that feature, it might be best not to output sequence numbers.

File: tools/generator.js [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+            'source': path,
+            'tags': tags,
+            'title': title,
+            'sequenceNumber': sequence,
+        };
+    },
+
+    _create_image: function (path, mime_type) {
+        let license = 'CC-BY 3.0';
+        let width = 0;
```
Comment (ptomato) [inline tools/generator.js]: Maybe something for a subsequent iteration, but I think we could read the image files to determine the width and height.

File: tools/generator.js [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+            'contentType': mime_type,
+            'source': path,
+            'tags': tags,
+            'title': title,
+            'sequenceNumber': sequence,
+        };
+    },
+
+    _create_image: function (path, mime_type) {
+        let license = 'CC-BY 3.0';
```
Comment (ptomato) [inline tools/generator.js]: What are your plans for this?

File: tools/generator.js [hunk 7]
```diff
@@ -0,0 +1,163 @@
+const Lang = imports.lang;
+const System = imports.system;
+
+const Gio = imports.gi.Gio;
+
+const Generator = new Lang.Class({
```
Comment (ptomato) [inline tools/generator.js]: `Generator` is a built-in type in later JS versions, so best not to use this name.

File: tools/generator.js [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    _create_set: function (path, title, child_tags, tags, thumbnail_path) {
+        let _tags = ['EknSetObject'];
+        let featured = false;
+
+        if (tags.length === 0) {
+            _tags.push('EknHomePageTag');
+            featured = true;
+        }
+
+        tags.forEach(tag => {
```
Comment (ptomato) [inline tools/generator.js]: `_tags.concat(tags)` or `_tags = [..._tags, ...tags]`

File: tools/generator.js [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    },
+});
+
+const USAGE = [
+    'usage: basin-generator <path_to_input_directory> <path_to_output_json>',
+    '',
+    'Utility for generating basin JSON files.',
+].join('\n');
+
+let main = function () {
```
Comment (ptomato) [inline tools/generator.js]: `function main() {` since otherwise it'll be an anonymous function assigned to the `main` variable

File: tools/generator.js [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+    _init: function (dir_path) {
+        this._dir_path = dir_path;
+        this._sequence = 0;
+    },
+
+    _create_root: function () {
+        return {
+            'articles': [],
+            'images': [],
+            'sets': []
```
Comment (ptomato) [inline tools/generator.js]: nitpick: trailing commas everywhere

File: basin-helper [hunk 11]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+PATH_TO_JSON=/tmp/basin.json
+PATH_TO_FILES=$1
+PATH_TO_SHARD=$2
```
Comment (ptomato) [inline basin-helper]: Maybe something for another iteration, but how about taking in an app ID and automatically finding the path to the shard?

File: configure.ac [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+AC_CACHE_SAVE
+
+# Required libraries
+# ------------------
+EOS_CHECK_GJS_GIR([Gio], [2.0])
+EOS_CHECK_GJS_GIR([GLib], [2.0])
+EOS_CHECK_GJS_GIR([EosShard], [0])
+EOS_CHECK_GJS_GIR([Xapian], [1.0])
+
+AC_CACHE_SAVE
```
Comment (ptomato) [inline configure.ac]: I would also recommend using `AX_CHECK_GIR_SYMBOLS_GJS` to make sure `Xapian.Document.add_numeric_value` is present.

File: basin-helper [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+
+echo $PATH_TO_FILES $PATH_TO_SHARD
+
+basin-generator $PATH_TO_FILES $PATH_TO_JSON
+basin $PATH_TO_JSON $PATH_TO_SHARD/output.shard
+eminem regenerate $PATH_TO_SHARD
```
Comment (ptomato) [inline basin-helper]: This is more a problem with eminem, but if manifest.json is a symlink to a manifest.json from a system-installed flatpak, then this will fail.

**Review by ptomato — CHANGES_REQUESTED** [review]
A few comments, plus errors that I ran into while testing this plus the other pull request.

File: src/packer.js (hunk 2 above)
Comment (tchx84) [reply src/packer.js]: Kind of, it was a quick hack to avoid having to rely on the set tittle (since it can be repeated e.g. in a hierarchy). But,I will fix that later.

File: tools/generator.js (hunk 6 above)
Comment (tchx84) [reply tools/generator.js]: I think we could define some convention to specify this globally (e.g. having just 1 file with the license string for ALL of the contents) OR have per-file convention  (e.g. something in the name of the file or even having a separate file like article.pdf.license). Need to think more about that.

For now, I will just leave it empty (e.g. "").

File: tools/generator.js (hunk 5 above)
Comment (tchx84) [reply tools/generator.js]: Yes, I was thinking to use pixbuf or something else, but even better if is something could also work for video/audio. But I leave this for later.

File: tools/generator.js (hunk 4 above)
Comment (tchx84) [reply tools/generator.js]: Yes, I thought about having that name convention as well, but for now I don't think it does any harm having that sequence number as it is (in both cases the order is random and at least it makes it easier to edit on the generated json file in case someone wants to enforce an specific order).

File: basin-helper (hunk 11 above)
Comment (tchx84) [reply basin-helper]: Not entirely sure I understand this. I don't think we need to enforce any path based on the app_id, the shards we create doesn't necessarily need to belong to an specific app, and we can try these shards with any app using the custom-content option I am adding to the e-k-l.

File: basin-helper (hunk 13 above)
Comment (tchx84) [reply basin-helper]: Maybe related my previous comment, but we don't need mix these two directories, since we will be able to use any custom shard with the custom-content option in e-k-l.

**Comment by tchx84** [comment]
@ptomato Ill create a follow up issue to address these details we discussed.

File: tools/processor.js [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+    'Utility that processes a directory to create the JSON input file for basin.',
+].join('\n');
+
+function main () {
+    let argv = ARGV.slice();
+
+    if (argv.length !== 2)
+        fail_with_message(USAGE);
+
+    let dir_path = argv[0];
```
Comment (ptomato) [inline tools/processor.js]: `let [dir_path, json_path] = argv`?

File: tools/processor.js [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+    let json_path = argv[1];
+
+    let processor = new Processor(dir_path, json_path);
+    let output_dict  = processor.run();
+    let output_json = JSON.stringify(output_dict);
```
Comment (ptomato) [inline tools/processor.js]: I suggest pretty printing the output so it's in a more editable form to begin with: `JSON.stringify(output_dict, null, '\t')`

File: basin-helper (hunk 11 above)
Comment (ptomato) [reply basin-helper]: I asked because you run `eminem regenerate` on this path, so to me it seems it can only be in the `~/.local/share/com.endlessm.subscriptions` area?

**Review by ptomato — CHANGES_REQUESTED** [review]
Maybe I need some clarification on the `eminem regenerate` thing.

File: basin-helper (hunk 11 above)
Comment (tchx84) [reply basin-helper]: No, it can be anywhere, you can generate shards wherever you please and then test them with --content-path= option.

EDIT: we could also provide a helper for doing this (or extend this tool to also do that), but I think we want to have this flexibility of creating shards/content anywhere we want (not restricting it to the directory convention, considering we will be able to load any directory with the content-path= option in e-k-l).

File: basin-helper [hunk 16]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+PATH_TO_JSON=/tmp/basin.json
+PATH_TO_FILES=$1
+PATH_TO_SHARD=$2
+
+echo $PATH_TO_FILES $PATH_TO_SHARD
+
+basin-processor $PATH_TO_FILES $PATH_TO_JSON
```
Comment (ptomato) [inline basin-helper]: In order to make this a bit more friendly you should abort if the wrong number of arguments are passed, and also abort if one of these tools exits with a nonzero exit code.

**Review by ptomato — CHANGES_REQUESTED** [review]
Just a minor comment on the helper script, it's a bit overwhelming when you run it without arguments.

**Review by ptomato — APPROVED** [review]
**CR pass**
